### PR TITLE
feat(base): change base to SeaweedFS

### PR DESCRIPTION
**Pull request by andrewazores**
- chore(base): change base to SeaweedFS
- add customized config template and entrypoint script
- add timed rebuilds

### CODE REVIEW

**Comment by tthvo** [comment]
Just curious here but wondering what are the reasons to switch to seaweed here?

**Comment by andrewazores** [comment]
A couple of things:
- SeaweedFS is Apache licensed, same as our projects, which is more comfortable than Minio's AGPL
- Configuration for SeaweedFS authentication/authorization is a bit simpler and easier. Should be feasible to hide it completely behind an auth proxy and run it with no auth on its own, though this at least breaks the presigned URL feature that we're currently using in 3.0, so that would need to be redesigned
- (supposedly) better performance and data consistency guarantees

**Comment by tthvo** [comment]
Oh i see! Thanks for explaining!!

>  though this at least breaks the presigned URL feature that we're currently using in 3.0, so that would need to be redesigned

I am pretty curious how this issue actually is going to work out.  But make sense to me.

**Comment by andrewazores** [comment]
In 3.0 when the client asks to download a file, Cryostat puts the requested data into the S3 provider (seaweed) if necessary (ie for an active recording. For an archived recording it's already there so we skip this first step), then responds to the client with a redirect response to another endpoint. When the client hits that endpoint, Cryostat asks S3 to generate a presigned URL for downloading that file, which comes with some query parameters and things that ensure the request has not been tampered with and the client is only able to access that one file. Cryostat then responds to the client again with another redirect response, this time pointing it to the S3 presigned URL. The client finally goes and makes a request to this URL, including the query parameters, and S3 sends the file back to the client.

In 2.2 or 2.3 I implemented a system where the client requests to download a particular file and the server generates a JWT and returns that back to the client, and the client can then make a new request to a different URL and pass that JWT as a query parameter instead of passing an Authorization header. The JWT is signed and encrypted and embeds information about the resource being requested etc., so when Cryostat receives a request with a JWT and no Authorization header, it's able to check that the signature is still valid, decrypt the token, and validate that the data embedded in the JWT matches the request it was attached to. If it does then Cryostat sends back the file the client requested.

So if 3.0 can't use presigned URLs in the end due to the auth proxy architecture, then it'll just need to reimplement something like that 2.x JWT download flow. From the client's perspective they are actually quite similar already, so this is just internal implementation detail.

**Comment by tthvo** [comment]
Understood now! Thanks for the detailed explanations!

**Comment by andrewazores** [comment]
I still need to get around to deploying 3.0 with Seaweed in OpenShift and an auth proxy, set up the Services and Routes, and test how authentication between Cryostat and Seaweed works, and then how authentication between Seaweed and an external client (your browser) works.

- we want presigned URL support, but this requires Seaweed to be configured with user accounts (not no-auth), like I added as a template config file in this PR
- but if Seaweed requires authentication, are we able to pass those credentials through the auth proxy along with the credentials required for the auth proxy itself? ie does Cryostat need to somehow supply both OpenShift credentials to the proxy, and have the forwarded request also include Seaweed credentials?
- we could probably avoid that requirement by exposing Seaweed through the proxy and a Route externally, and directly (no proxy) on a Service internally, but it would be even better security to always pass traffic through the proxy
- we could also avoid that by not using an auth proxy in front of Seaweed at all and simply depending on strong generated secrets as in this PR's template config. This is not ideal since Seaweed is exposed directly internally and externally, but we can restrict that to a single port and with strong secrets for credentials

**Comment by tthvo** [comment]
> though this at least breaks the presigned URL feature that we're currently using in 3.0, so that would need to be redesigned
> how authentication between Seaweed and an external client (your browser) works.

Actually, I am wondering if the client needs to authenticate with Seaweed at all if using presigned url. Wouldn't it be a publicly available url to download anonymously? Then, I guess seaweed can just be behind proxy with OpenShift credentials. Or might be missing sth.

**Comment by andrewazores** [comment]
The client won't need Seaweed credentials when using a presigned URL, but it will still need to pass through the auth proxy with OpenShift credentials. That should be easy, I don't see any reason that won't look basically the same as how it already looks with oauth2-proxy in smoketest.

**Comment by tthvo** [comment]
Sounds good! Thanks for the details ^^

**Comment by andrewazores** [comment]
@ebaron I think there's more work to do in here around tuning the volume server parameters - for some reason it didn't like the defaults of "no max (number of) volumes" and 1GB per volume, so over in https://github.com/cryostatio/cryostat-helm/pull/111 I have it deployed with some smaller hardcoded limits that seem to work for a basic smoketest.

Otherwise, I think this is settled enough to use as a launching point for more work like https://github.com/cryostatio/cryostat-helm/pull/111 , https://github.com/cryostatio/cryostat3/pull/241 , https://github.com/cryostatio/cryostat-operator/issues/710 .

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,18 @@
-FROM docker.io/minio/minio:RELEASE.2023-12-14T18-51-57Z
+ARG builder_version=8.9
+ARG runner_version=8.9
 
-ENV MINIO_USER minio
-ENV MINIO_UID_GID 5001
+FROM registry.access.redhat.com/ubi8/ubi:${builder_version} AS builder
+ARG ref=master
```
[the remark this answers is not in the document]
Comment (andrewazores) [reply Dockerfile:5]: I think for our own release branches, or whatever release strategy we use on this repo, that will be the right play. Maybe in the release branch we update this file to hardcode in a specific ref (git tag) from SeaweedFS.

For our own `main` branch I tried to set up something similar to what we have for `cryostat-db` - that container layers on top of the latest `sclorg/postgres` container and rebuilds once per week, so here I track the `master` branch and have the CI set up to rebuild once per week, too.

[the remark this answers is not in the document]
Comment (ebaron) [reply Dockerfile:5]: Seems like they release weekly. If we could add something in CI to find the latest release tag and build from that, that would be probably be safer. That can be an RFE though.

**Review by ebaron — APPROVED** [review]
Looks good!

**Pull request merged**